File: src/mkcpp.cpp
/**************************************************************
 * File Name   : mkcpp.cpp
 * Author      : ThreeDog
 * Date        : Thu Nov  3 11:05:14 2016
 * Description : 快捷命令 制作.cpp文件
 *
 **************************************************************/


#include <iostream>
#include <cstdlib>
#include <cstring>
#include <vector>
#include <fstream>

using namespace std;

int main(int argc,char **argv){

	int size = 20;

	ofstream output_file(argv[1],ios::trunc);
	if(argc < 2){
		cerr << "arguement error: 命令mkcpp需要包含一个参数 " << endl;
		return 1;
	}
	else if(argc > 2){
		cerr << "arguement error: 命令mkcpp只能包含一个参数 " << endl;
		return 1;
	}

	if(!output_file){
		cerr << "open file " << argv[1] << "fail" << endl;
		return 3;
	}
	time_t now_time;
	now_time = time(NULL);
	output_file << "/**************************************************************"<<endl;
	output_file << " * File Name   : " << argv[1] << endl;
	output_file << " * Author      : ThreeDog "<<endl;
	output_file << " * Date        : " << ctime(&now_time);
	output_file << " * Description : " << endl << " *" << endl;
	output_file << " **************************************************************/"<<endl;

	output_file << endl;
	output_file << "#include <iostream>" << endl;
	output_file << endl;
	output_file << "using namespace std;" << endl;
	output_file << endl;
	output_file << "int main (int argc, char **argv){" << endl;
	output_file << endl;
	output_file << endl;
	output_file << endl;
	output_file << endl;
	output_file << "return 0;" << endl;
	output_file << "}" << endl;

	char *a = new char[size];
	char *name = new char[size];
	strcpy(name,argv[1]);
	strcpy(a,"vi ");
	char *n = strtok(argv[1],".");

	while(strlen(n) > size){
		char *r = new char[size+10];
		strcpy(r,a);
		delete[] a;
		a = r;
		size += 10;
		}

	strcat(a,name);
	system(a);

	return 0;
}

File: src/golang/mkcpp.go
/**************************************************************
* File Name   : mkcpp.go
* Author      : ThreeDog
* Date        : 2019-12-23 20:02:00
* Description : 
*
**************************************************************/

package main

import (
	"fmt"
	"os"
	"log"
	"bytes"
	"time"
)

func PathExists(path string) (bool, error) {
	_, err := os.Stat(path)
	if err == nil {
		return true, nil
	}
	if os.IsNotExist(err) {
		return false, nil
	}
	return false, err
}

/**
 检查错误
 */
func checkErr(err error){
	if err!=nil{
		log.Fatal(err)
	}
}

func main()  {
	// 命令行参数获取
	if len(os.Args) != 2 {
		fmt.Printf("arguement error: 命令mkc需要有且只有一个参数 \n")
		return 
	}
	
	res,_ := PathExists(os.Args[1])
	// 如果文件已经存在，替换一下文件名 
	if res {
		fmt.Printf("文件%s已存在，是否覆盖创建？（旧文件将被重命名为：%s.bak ）Y/N ",os.Args[1],os.Args[1])
		var ok string
		fmt.Scanln(&ok)
		if ok == "Y"{
			err := os.Rename(os.Args[1],os.Args[1]+".bak")
			if err != nil{
				fmt.Printf("文件%q已存在，将文件重命名为%q.bak时失败 \n",os.Args[1],os.Args[1])
				return 
			}
		}else{
			return 
		}
	}
	
	// 打开文件
	file,err:=os.OpenFile(os.Args[1],os.O_CREATE|os.O_APPEND|os.O_WRONLY,0644)
	checkErr(err)
	var buffer bytes.Buffer
	buffer.WriteString("/**************************************************************\n")
	buffer.WriteString("* File Name   : " + os.Args[1] + "\n")
	buffer.WriteString("* Author      : ThreeDog\n")
	buffer.WriteString("* Date        : "+time.Now().Format("2006-01-02 15:04:05")+"\n")
	buffer.WriteString("* Description : \n")
	buffer.WriteString("*\n")
	buffer.WriteString("**************************************************************/\n\n")
	buffer.WriteString("#include <iostream>\n\n")
	buffer.WriteString("using namespace std;\n\n")
	buffer.WriteString("int main (int argc , char** argv)\n")
	buffer.WriteString("{\n")
	buffer.WriteString("\treturn 0;\n")
	buffer.WriteString("}\n")

	file.Write([]byte(buffer.String()))
	file.Close()
}
